LUN-132: Refactor documents fetching logic and route parameters

klinzo opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

This paves the way for `LUN-132: Favorite word list`. I had to refactor the documents fetching logic and the route params to be able to pass custom documents belonging to different disciplines to the exercises. Since the requirement only changed during development (previously all exercises should be possible for favorites) and because I think the new approach makes sense I refactored this for all exercises.

klinzo commented on src/routes/choice-exercises/components/SingleChoiceExercise.tsx:
```diff
@@ -153,47 +144,40 @@ const ChoiceExerciseScreen = ({
-      />
-
-      <ServerResponseHandler error={error} loading={loading} refresh={refresh}>
-        {documents && correctAnswer && currentDocument && (
```
Just removed this conditional as it was not needed anymore.

SteffiStuffel commented on src/routes/write-exercise/__tests__/WriteExerciseScreen.spec.tsx:
```diff
@@ -1,89 +0,0 @@
-    }
-  }
-
-  it('should show network error', async () => {
```
I see why you deleted this tests, but can you maybe add - in a more generic way - a test (ServerResponseHandler.spec.tsx)

SteffiStuffel commented on src/routes/write-exercise/components/InteractionSection.tsx:
```diff
@@ -135,7 +133,7 @@ const InteractionSection = (props: InteractionSectionProps): ReactElement => {
       <Speaker>
         <AudioPlayer
-          document={documentWithResult}
+          document={documentWithResult.document}
```
just `document` should also be fine, as you destructured it above?

SteffiStuffel commented on src/routes/ExercisesScreens.tsx:
```diff
@@ -24,17 +25,24 @@ interface ExercisesScreenProps {
-    if (item.title === labels.exercises.wordChoice.title && discipline.numberOfChildren < MIN_WORDS) {
-      Alert.alert(labels.exercises.wordChoice.errorWrongModuleSize)
-    } else {
+    if (documents) {
```
Oke I see you changed the behaviour. I'm fine with this, also may help when adding the unlocking stuff.

klinzo replied:
I removed this because I thought we need it to be able to do the exercise for favorites if there are less then three favorites. Since this is not required anymore I could revert it, but I think this solution makes more sense/is easier. But I am open to reverting if I should

SteffiStuffel replied:
Keep it as it is, it's fine or even better.

SteffiStuffel reviewed (changes requested):
Really nice refactoring. I like it.

While testing I found some bugs:
- Navigation Animation is broken (see detailed explanation above)
- Retry hole exercise is broken, after retrying only wrong results. Steps to reproduce:
 1. Open writing exercise
 2. Make some answers correct and some incorrect
 3. Finish exercise and retry only wrong words
 4. Make all answers incorrect
 5. Click on the ExerciseFinishedScreen on Retry Exercise
 6. Exercise and Header is broken

Wordlist shows wrong words. Steps to reproduce:
1) Go to write exercise
2) Finish exercise with some words correct and some words incorrect
3) Retry wrong words and do all of the wrong
4) Finish exercise and close it
5) Go to word-list (or exercise one or two), only the previously wrongly done words are show. 
Expected behaviour: all words should be shown.

klinzo commented on src/routes/ResultDetailScreen.tsx:
```diff
@@ -54,41 +53,40 @@ const ResultDetailScreen = ({ route, navigation }: ResultScreenProps): JSX.Eleme
-    navigation.navigate('WriteExercise', {
-      discipline: { ...result.discipline },
-      retryData: { data: entries }
+    navigation.push('WriteExercise', {
```
Reverted, no idea why I changed it.

klinzo commented:
> Navigation Animation is broken (see detailed explanation above)

Reverted, no idea why I changed it.

> Retry hole exercise is broken, after retrying only wrong results. Steps to reproduce:

Should be fixed.

> Wordlist shows wrong words. Steps to reproduce:

Should be fixed.

> Wrong order in Word-list:

Should be fixed, was because of copy by reference instead of value.

SteffiStuffel commented:
Tested again. Works fine now.

SteffiStuffel reviewed (changes requested):
Merge blocked till next release.
